python/ray/_private/worker.py

```diff
@@ -1977,6 +1990,7 @@ def sigterm_handler(signum, frame):
 
 @PublicAPI
 @client_mode_hook
+@with_connect_or_shutdown_lock
```

Comment:
Making sure all the Ray public APIs are thread safe is a big follow-up, but this looks fine to me.